lib/build.ts

```diff
@@ -187,3 +194,169 @@ export async function runBuildTask(
+//   for the best.  It has to do this in order to maintain backwards-compatible behavior:
+//
+//      + V2 manifests for all images: If using multiarch images, Docker will
+//        infer the platform from where the build engine is running.  If using single arch
+//        images, Docker will assume the declared image arch is the correct platform.
+//          -> Result: Docker builds, with a warning if the images do not match the
+//                     current platform.
+//
+//      + Some V1 manifests:
+//          -> Result: Docker builds, assuming that the platform matches where the build
+//                     is occurring.
+//
+//      NOTE!  Even when a build succeeds, if Docker is making assumptions then it is possible that
+//      the image will cause an execution error on the target platform if the architecture is not
+//      correct.
+//
+//   So what do we do?
+//
+//      + In the case of all images having a V2 manifest, we can simply pass the platform flag.
+//
+//      + In the case of having some V1 manifests ...
+//          - If we have balenalib images, we know that they have been updated.  Throw an error
+//            and tell the user to update
+//          - If we do not have balenalib images, don't pass the platform flag.  Let Docker
+//            close its eyes and hope for the best.  This opens the possibility for an exec
+//            error, but also allows users to continue to use v1 images if they want to / need to.
+//
+async function checkAllowDockerPlatformHandling(
+	task: BuildTask,
+	docker: Dockerode,
+): Promise<boolean> {
+	let imageReferences: string[];
+
+	const checkHasV1Manifest = async (repository: string) => {
+		try {
+			const manifest = await getManifest(docker.modem, repository);
+			return manifest.Descriptor.mediaType === MEDIATYPE_MANIFEST_V1;
+		} catch {
+			// eat the exception... yummy!
+		}
+
+		// at this time, treat any errors as true (v1 manifest)
+		return true;
+	};
+
+	if (task.imageName) {
+		imageReferences = [task.imageName + !!task.tag ? `:${task.tag}` : ''];
+	} else {
+		if (!task.dockerfile) {
+			// Sanity check
+			console.warn('Build task does not have an associated Dockerfile');
+			return false;
+		}
+		const { DockerfileParser } = await import('dockerfile-ast');
+		const parsedDockerfile = DockerfileParser.parse(task.dockerfile);
+		const dockerInstructions = parsedDockerfile.getInstructions();
+		const fromInstructions = dockerInstructions.filter(
+			(inst) => inst.getKeyword() === 'FROM',
+		);
+		if (fromInstructions.length === 0) {
+			// Sanity check
+			console.warn(
+				'Build task associated Dockerfile does no reference any images',
+			);
+			return false;
+		}
+		imageReferences = [];
+		fromInstructions.forEach((inst) => {
+			const fromLineArguments = inst.getArguments();
+			if (fromLineArguments.length === 0) {
+				// bail out, no image specified?
+				return;
+			}
+
+			imageReferences.push(fromLineArguments[0].getValue());
```

Comment:
Do we really have to specify the index here? I think this will break if additional args are provided in the FROM line.
```
FROM --platform=arm/v7 alpine AS build
```
I think the above Dockerfile will result in us passing `--platform=arm/v7` to the list of images, but I could be mistaken.

Reply:
https://docs.docker.com/engine/reference/builder/#from

Reply:
Well shoot... this presents two problems

1. There doesn't appear to be a good library for parsing the FROM line (at least, as far as my looking around is concerned, I haven't found one).  But I can still do a (hacky?) look through the parameters and make sure I am actually getting the image.  This doesn't seem like a terrible option.   
2. It did not occur to me previously that a user might specify `platform` here.  How should we handle this?   Going to [FD to discuss](https://www.flowdock.com/app/rulemotion/i-cli/threads/RuSu1KiWOn62xaGy7O2sn8m8BUc)

Reply:
Talking with @klutchell on FD, `platform` on the command line is ignored when it is specified in the dockerfile.  So we will just let the engine decide what to do, ignoring it in the dockerfile.  (regarding 2nd point)